### PR TITLE
WIP: Feature: Implement forwardRef in wp.element.createHigherOrderComponent

### PR DESCRIPTION
**Pull request by nerrad**
## Description

Using [React.forwardRef](https://reactjs.org/docs/forwarding-refs.html) it's possible to forward refs to children components from their parents.  This pull seeks to implement this with `createHigherOrderComponent` so any components wrapped by a HOC created through that will automatically have the ref passed through the HOC to them.

Currently this is a work in progress and it's not working as expected:

- I _expected_ that I'd be able to just return the result from `forwardRef(forwardedRefComponent)` but doing a build with that results in errors in the console when executing the GB editor.  Doing `forwardRef(forwardedRefComponent).render` appears to work well for the built files but there are failing tests so I suspect there's some issues not seen in basic user testing of the editor.
- The test I added for verifying the ref is actually forwarded isn't passing so it appears that the isn't actually getting forwarded.

I'm putting this pull up just so others can see the work so far and it might be quickly apparent what needs fixed or whether this is only feasible with some refactoring of other things.

Related issues/comments:

- https://github.com/WordPress/gutenberg/pull/6261#discussion_r191755736
- https://github.com/WordPress/gutenberg/pull/6480#issuecomment-392805451

See [summary comment](https://github.com/WordPress/gutenberg/pull/7557#issuecomment-401794634) regarding findings.  Basically we need a decision on whether to proceed (and how to proceed) with finishing off this pull considering the findings or whether individual HOCs should be responsible for implementing `forwardRef`.

## How has this been tested?

* [ ] Trying various blocks in the editor
* [x] Added unit tests.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 

## Rollout

Here's some things needing done in this branch before its ready for merge.

* [x] All tests that fail in this branch (because of lack of enzyme support) switched to use alternative test utilities in pulls off master (see pulls linked through this issue).
* [ ] Master merged in and unit-tests passing.
* [ ] Jest updated so that snapshots can be updated (currently they don't print the pretty display name because the current Jest version doesn't detect forwardRef wrapped components properly).
* [ ] Making any necessary changes to existing HOCS implementing forwardRef manually.

### CODE REVIEW

**Comment by nerrad** [comment]
It looks like enzyme support for forwardRef might be coming with [this pull](https://github.com/airbnb/enzyme/pull/1592)

**Comment by gziolo** [comment]
@nerrad, I would relay on `enzyme` anymore, they are very slow with fixing changes introduced with 16.3 and up. They didn't release any new version since January or something like that. We are leaning towards using `react-test-renderer`. See this commit from @youknowriad https://github.com/WordPress/gutenberg/pull/7527/commits/cc95059071469c56a330444d71a9167f241b393e.

**Comment by nerrad** [comment]
I can dig dropping enzyme, but the issue here is that all the failing tests are existing tests using enzyme.  So it sounds like all those tests would have to be replaced with react-test-renderer.  Should that be done in a separate pull?

**Comment by nerrad** [comment]
@aduth I'm a bit stuck with how we want to forward Refs through the HOCs as a part of this pull and I'm not sure yet I'm on the right approach (or even IF it's a good approach).  Currently with the approach I've taken in here, the leaf components would be able to check for a `forwardedRef` prop and use that for their ref.  There's some niggling questions though:

1. What about existing HOCs like `withGlobalEvents` which creates its own ref on the wrapped component?
2. Do we convert existing leaf components (such as what you did here https://github.com/WordPress/gutenberg/pull/6527) to use the new `forwardRef` if it exists?
3. Is this just a higher level extraction so any HOC's using `createHigherOrderComponent` don't have to do the `forwardRef` call and can just use `forwardRef` to assign a `ref` on `WrappedComponents`?

**Comment by nerrad** [comment]
I’ve wrestled with this a bit and changed things a bit and tweaked the unit-test that demonstrate one possible use case.  It looks like it’d be a bit better than what I originally had, but there’s still some problems.

- When using with compose, the “leaf” component would have to use `forwardedRef` prop directly as opposed to being implemented as the return from `forwardRef` wrapping it.
- `createHigherOrderComponent` is going to always return a `forwardRef` symbol.  This means that depending on where `pure` is used in the compose stack, it’s not going to work as expected (and definitely is harder - not sure yet how) to assert if state changes result in rerender.

I’m beginning to wonder whether its a good idea to do forwardRef here and instead just leave it to individual HOC elements to configure.  Part of the problem I think is that there’s not really direct access (that I can see) to the passed in component that the “OriginalComponent” is mapped to.

**Comment by nerrad** [comment]
If the current approach is okay, one thing we maybe could do is have `createHigherOrderComponent` accept a flag to indicate whether it does `forwardRef` or not ( maybe `doForwardRef` defaulting to true)  and if it isn't then it skips that logic.  Then the logic for `pure` would set that flag to false.  

So something like this:

```jsx
export function createHigherOrderComponent( doForwardRef = true ) {
	return ( mapComponentToEnhancedComponent, modifierName ) => {
		return ( OriginalComponent ) => {
			const WrappedComponent = mapComponentToEnhancedComponent(
				OriginalComponent );
			const EnhancedComponent = forwardRef(
				( props, ref ) => {
					const forwardedRef = ref !== null ? ref : props.forwardedRef;
					return <WrappedComponent
						{ ...props }
						forwardedRef={ forwardedRef }
					/>;
				}
			);
			const { displayName = OriginalComponent.name || 'Component' } = OriginalComponent;
			EnhancedComponent.displayName
				= `${ upperFirst( camelCase( modifierName ) ) }(${ displayName })`;
			return EnhancedComponent;
		};
	};
}


export const pure = createHigherOrderComponent( false )( ( Wrapped ) => {
	if ( Wrapped.prototype instanceof Component ) {
		return class extends Wrapped {
			shouldComponentUpdate( nextProps, nextState ) {
				return ! isShallowEqual( nextProps, this.props ) || ! isShallowEqual( nextState, this.state );
			}
		};
	}

	return class extends Component {
		shouldComponentUpdate( nextProps ) {
			return ! isShallowEqual( nextProps, this.props );
		}

		render() {
			return <Wrapped { ...this.props } />;
		}
	};
}, 'pure' );
```

**Comment by nerrad** [comment]
Admittedly, I'm not versed in javascript symbols.  So I'll do a bit of reading on it as well.

**Comment by nerrad** [comment]
If we _do_ go with the current approach (and work out some solution for testing `pure`) then leaf components could probably have their definition be something like this:

```jsx
export const Button = ( props ) => {
    return <Button ref={ props.forwardedRef } />
};

export const ButtonWithRef = forwardRef( ( props, ref ) => {
    const forwardedRef = ref !== null ? ref : props.forwardedRef
    return <Button ref= { forwardedRef } />
};
```

**Comment by nerrad** [comment]
I'm wary about using the idea I had for adding an argument to the existing `createHigherOrderComponent` method for conditionally applying `forwardRef`.   I think it'd be best if we do a new method.:

> When you start using forwardRef in a component library, you should treat it as a breaking change and release a new major version of your library. This is because your library likely has an observably different behavior (such as what refs get assigned to, and what types are exported), and this can break apps and other libraries that depend on the old behavior.

> Conditionally applying React.forwardRef when it exists is also not recommended for the same reasons: it changes how your library behaves and can break your users’ apps when they upgrade React itself.

**Comment by nerrad** [comment]
With https://github.com/WordPress/gutenberg/pull/7557/commits/12ad2bca1f9931477a1739afa8a1479c96dff852 I finally figured out how to extract the WrappedComponent instance for asserting changing state on a Component instance wrapped by `pure` results in a re-render.  This successfully converts the previous `pure` tests from enzyme to `react-test-renderer`.

**Comment by nerrad** [comment]
In https://github.com/WordPress/gutenberg/pull/7557/commits/ed0e374571613ac5fb3fa1466f82677a50de50c3 I've added a test that demonstrates my concern with `pure` and the changes in this branch.  As noted, depending on the order pure is used within `compose`, it will handle a class component as expected because it ends up getting passed in to pure as a `forwardRef` wrapped function component.

**Comment by nerrad** [comment]
So it appears this behaviour of pure is something that already exists in master (obviously).  Since I'm not sure whether its "known" or expected behaviour I duplicated the test I added here in a branch of master and created a pull (#7654).  That way if that's determined to be invalid, then the same concern I have here might be considered invalid as well.

**Comment by nerrad** [comment]
## What next for this pull?

A summary of some of the things uncovered in this work and what needs a decision (with some potential options).

### Implementing `forwardRef` in `createHigherOrderComponent` changes return type.

`forwardRef` on a component returns essentially a functional component.  This affects any code expecting the same type of component returned as what was passed in.  While in most cases this might not be problematic, it is technically a breaking change.   In the case of `pure` this is noticeable when `pure` is used as a part of a composed HOC (i.e. `withSelect` or `withDispatch`) IF the component being composed is a class component with state.  

**Things for consideration**:

- change `pure` so it is only  HOC that makes functional components "pure" (thus shallow compare on props only) and implement the expectation that if class components need to be pure they should extend `React.PureComponent`.  This immediately makes expectations clear and removes the potential for gotchas with using of compose.  Note: `recompose` has a `pure` HOC that only applies to functional components (shallow compare on props only).
- documentation will need to be clear about what `createHigherOrderComponent` does (not only composing the higher order component itself but also applying `forwardRef`.

### Enzyme does not support `forwardRef`.

This will mean that there's quite a few tests that will need converted to `react-test-renderer`.  Totally doable and might be a bit better in the long run as `react-test-renderer` is more low level.  However it does make writing tests more difficult.

### Introduction of `forwardedRef` prop on all HOCs created with `createHigherOrderComponent`

This means that the composed classes will need to pass through `ref` (i.e. `withGlobalEvents`) using the exposed `forwardedRef` component.  Otherwise "leaf" components will need look for the `forwardedRef` prop.

### Inconsistency among HOCs in GB

Currently there's still a few HOCs that don't use `createHigherOrderComponent`.  To be useful, it'd be good to have all the HOC's go through this function.  This also means existing HOC's that implement forwardRef individually will need to be updated to implement `forwardedRef` prop.

**Comment by aduth** [comment]
Thanks for spending the time working through this @nerrad .

One consideration is an extreme idea that's been proposed in passing on a few occasions: Not exposing `Component` from `@wordpress/element`, and instead putting full weight into higher-order components as the interface to define enhanced component behaviors. The thought here is to better guarantee future compatibility with e.g. breaking changes of deprecated `componentWillMount` and other lifecycle methods ([source](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html)). As it impacts this pull request, if we didn't need to consider `pure` as operating on state and instead only on props (or general not considering `instanceof Component`), it shouldn't be as much of an issue? See also [`withState`](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-state).

I don't love the idea of allowing a higher-order component to opt-out of ref forwarding, since it makes it less obvious for a developer to know what will happen when they assign a `ref`. If we can't forward for all, I don't think we should forward for any.

**Comment by nerrad** [comment]
Re not exposing component:  In principle I understand that.  However practically, I wonder how much benefit that will really bring.  Will GB still be "compatible" with custom components written in react by third parties?  Will it narrow the use-case for GB components to a WP (or @wordpress package) context?  I'm wary of this being a way to make it easier to bridge future compatibility issues with React lifecycles.  Would this also mean dropping the [use of `React.StrictMode`](https://github.com/WordPress/gutenberg/blob/c7ce08beb514c1e4da5f428a03c07cb581d87703/edit-post/editor.js#L24).

> As it impacts this pull request, if we didn't need to consider pure as operating on state and instead only on props (or general not considering instanceof Component), it shouldn't be as much of an issue? See also withState.

I'm not sure changing `wp.element` so `Component` isn't exposed will affect that.  I think we can just decide that `pure` will only make functional components "pure" and then some other interface will be for handling state and props purity.  What about introducing a `withPureState`?  

> I don't love the idea of allowing a higher-order component to opt-out of ref forwarding, since it makes it less obvious for a developer to know what will happen when they assign a ref. If we can't forward for all, I don't think we should forward for any.

I concur and came to the same conclusion over the course of doing this ticket.

**Comment by nerrad** [comment]
Noting this comment from @gziolo in [slack](https://wordpress.slack.com/archives/C02QB2JS7/p1530714354000002):

> @nerrad, I’m a bit late, but I wanted to test performance regression properly. Regarding HOCs, my personal view is that all of them should be created using `createHigherOrderComponent` helper for consistency even without `ref` handling. At the moment it standardizes how you can discover those HOCs when using React devTools. Another issue as noted above is that when you compose HOCs you just need to ensure that all of them pass down refs properly. I’d encourage to continue your work and don’t worry as much about breaking changes. I don’t think there are many plugins that use this helper if any. As long as we can ensure that existing hooks like https://github.com/WordPress/gutenberg/blob/master/editor/hooks/custom-class-name.js#L54-L79 work, we can iterate.

**Comment by nerrad** [comment]
I think it might be a good first step to get all the GB HOCs using `createHigherOrderComponent` in a separate pull.  That way all tests etc on those components should be up-to-date and useful for catching any potential regressions as work is continued on this.  I'll try to carve out some time today to do that.

**Comment by nerrad** [comment]
see #7714 for updating all HOCs to use `createHigherOrderComponent`

**Comment by nerrad** [comment]
Note to self:  Investigate using a render prop in `createHigherOrderComponent` for maybe dealing with some of the shortcomings I've discovered so far.

**Comment by nerrad** [comment]
@aduth and/or @gziolo in https://github.com/WordPress/gutenberg/pull/7557/commits/58bd2d1ac0d33b8b6f8a0bc39c1b9b7fa0e3f7d6 and https://github.com/WordPress/gutenberg/pull/7557/commits/1b5612a86a1e848cdec64db666428e7eee8337eb I think it's pretty close to what the forwardRef implementation will look like in `createHigherOrderComponent` and tests are covering expectations.  I'm going to start reviewing all the failing tests with this change and update any necessary existing uses of refs in GB.  Part of this will involve ensuring all "leaf" components will conditionally set a ref similarly to how the test is doing it.  Namely, something like:

```jsx
const LeafComponent = forwardRef( ( props, ref ) => {
	const forwardedRef = props.forwardedRef || null;
	return <input type="text" ref={ ref || forwardedRef }/>;
} );
```
Any concerns/questions based on where things are at currently?

**Comment by nerrad** [comment]
The `core-blocks/test/full-contents.js` has failures that are kind of odd in relation to expectations described by fixtures.  Here's an example:

`core_latest-posts__displayPostDate.html` has this:

```
<!-- wp:core/latest-posts {"postsToShow":5,"displayPostDate":true} /-->
```

But the expectation for serialized html according to the `core__latest-posts__displayPostDate.serialized.html` fixture is:

```
<!-- wp:latest-posts {"displayPostDate":true} /-->
```

Tests run in this branch report this fail for that fixture:

screenshot of fail

So what I don't understand is I would expect the _fixture_ for serialized to be the same as the starting html (and thus this test would pass in this branch).  But it appears its not.  Doesn't make sense why the fixture would be different than the html in this case.

File: blocks/api/registration.js [hunk 1]
```diff
@@ -55,6 +57,15 @@ const POST_FORMAT_BLOCK_MAP = {
 	video: 'core/video',
 };
 
+const isEditPropValid = ( settings ) => {
```
Comment (aduth) [inline blocks/api/registration.js]: Two things:

- This seems like an enhancement we should support separate from `forwardRef` integration into `createHigherOrderComponent`, and could be its own pull request.
- This is definitely a piercing of the "element" abstraction. It seems like a function which should be offered by `@wordpress/element`. On glance, it appears similar to but not quite the same as what's already provided through `isValidElement`. Digging further, I found there is the [official `react-is` module](https://www.npmjs.com/package/react-is) which includes an `isValidElementType` method. We'd have to decide how we want to fit this into our abstraction. Honestly, I'm not sure why there's a `React.isValidElement` but not also the `isValidElementType` bundled; they seem appropriate together, and with that in mind I wouldn't mind if our package exported `isValidElementType` (though leveraging `react-is` underneath).

**Review by aduth — COMMENTED** [review]
Looking at the actual implementation in `createHigherOrderComponent`, it seems like the application of ref forwarding is an enhancement atop the enhanced component. In this vein, it seems as though it could be a higher-order component of its own (`withForwardedRef`). This could make it more straightforward for someone who's opting in to another higher-order component but wants to respect `ref` assignment. At the same time though, it requires the conscious foresight to understand that `ref` assignment is otherwise lost in a higher-order component wrapper.

Developer improvement considerations notwithstanding, I also worry about the performance impact. With these changes, each of our higher-order components becomes effectively double-wrapped, correct? Perhaps even triple wrapped, if you consider the third being the result of `forwardRef`.

**Comment by aduth** [comment]
FWIW, regardless of where this goes, the two separate enhancements identified are valuable discoveries in their own right (https://github.com/WordPress/gutenberg/pull/7557#discussion_r210355136, https://github.com/WordPress/gutenberg/pull/7557#discussion_r210352567).

**Comment by gziolo** [comment]
> Looking at the actual implementation in `createHigherOrderComponent`, it seems like the application of ref forwarding is an enhancement atop the enhanced component. In this vein, it seems as though it could be a higher-order component of its own (`withForwardedRef`). This could make it more straightforward for someone who's opting in to another higher-order component but wants to respect ref assignment. At the same time though, it requires the conscious foresight to understand that ref assignment is otherwise lost in a higher-order component wrapper.

This is a very interesting idea. It probably would have to be a wrapper on the compose itself:

```js
composeWithRef( [
    withHOC1,
    withHOC2,
    withHOC3,
    withHOC4,
] )( MyComponent )
```

which would be internally implemented as something like:
```js
const composeWithRef = ( HOCs) => compose( [
    ( OriginalComponent ) =>
        React.forwardRef( ( props, ref ) => <OriginalComponent { ...props } forwardedRef={ ref } /> ),
    ...HOCs,
    ( OriginalComponent ) =>
        ( { forwardedRef, ...props } ) => <OriginalComponent ref={ forwardedRef } { ...props } />,
] )( MyComponent )
```

**Comment by nerrad** [comment]
Hmm, ya a withForwardedRef component (or maybe better a variation of the `composeWithRef` idea @gziolo) is interesting.  I guess what route to take comes down to whether GB just wants ref to be exposed automatically on any wrapped leaf components (which was what sparked this pull I think?) or whether it is something that should be only implemented for specific composed components?  

The performance question is important and relevant to answering my previous question because wrapping _everything_ with `forwardref` just so the ref exposed might not be that performant and thus it could be better to just wrap as needed.

**Comment by nerrad** [comment]
> FWIW, regardless of where this goes, the two separate enhancements identified are valuable discoveries in their own right (#7557 (comment), #7557 (comment)).

Agreed, and I also agree they could be done as separate pulls (which I'm happy to do when I can carve out some time...hopefully this week).

**Comment by nerrad** [comment]
@aduth, @gziolo - it looks like `React.pure` will come with `forwardRef` support built-in in a [future version of react](https://github.com/facebook/react/pull/13822).  This seems to lend credence to the idea of automatically using forwardRef on HOC creation.  I realize GB has its own implementation of `pure` though.

**Review by gziolo — CHANGES_REQUESTED** [review]
Not that React hooks are about to land, we probably don't have to bother with `forwardRef` anymore. It also turns out that this isn't something we need to workaround in Gutenberg codebase.

Do you feel like this is something that still needs to be implemented? It definitely got stale and I'm going to add a proper label to make triaging PRs easier in the future.

In my opinion, we should either close it or refresh. The latter might be not efficient though given that React hooks should make HOCs legacy approach.

**Comment by nerrad** [comment]
I'm 👍 for closing.

**Pull request closed without merging**